### ISSUE TITLE
[unified-signatures] False positive on "required or omit" parameter with same return value

### ISSUE THREAD

**Issue by G-Rath** [typescript-eslint/typescript-eslint#961]
```ts
export function getUserRoles(): void;
export function getUserRoles(userId: number): void; // warning: this could be unified
export function getUserRoles(userId?: number): void {
}
```

While the above looks unifiable, there's a subtle difference: if `userId` is provided, it *must* be a number - i.e you can't explicitly pass `undefined`:
```ts
const fn = (): { p?: number } => {};

getUserRoles();
getUserRoles(1);
getUserRoles(undefined); // error - undefined is not assignable to type number!
getUserRoles(fn().p); // error - undefined | number is not assignable to type number!
```

If you unify function, the last two function call examples will be valid.

The use case is that the API I'm working w/ has two endpoints that both return `UserRoles[]`: `Users/Roles` & `Users/:userId/Roles`.

As such, if you provide a `userId`, it calls `Users/:userId/Roles`, otherwise it'll call `Users/Roles`.
This overload definition prevents you from being able to pass a value of `undefined`, as that'll return *all* roles when you expected the rules for a user.

i.e

```ts
async listUserRoles() {
  // user not loaded yet...
  return await getUserRoles(userId).map(role => (
    <ListItem>
      {role.name}
    </ListItem>
  ));
}
```

```JSON
{
  "rules": {
    "@typescript-eslint/unified-signatures": "warn"
  }
}
```

**Versions**

| package                            | version |
| ---------------------------------- | ------- |
| `@typescript-eslint/eslint-plugin` | `2.1.0` |
| `@typescript-eslint/parser`        | `2.1.0` |
| `TypeScript`                       | `3.6.2` |
| `ESLint`                           | `6.3.0` |
| `node`                             | `10.15.3` |
| `npm`                              | `6.11.3` |

**Comment by bradzacher** [comment]
I think this is just how the rule works.

Whilst there are subtle differences between the two differences, the thinking is that combining the parameters will make the devx simpler / code cleaner.
It's a conscious decision to be made when using this rule

Take the following example using "more concrete types":

```ts
declare function foo(a: number): void;
declare function foo(a: string): void;

foo('');                      // valid
foo(1);                       // valid
foo(('' as string | number)); // error

declare function bar(a: number | string): void;

bar('');                      // valid
bar(1);                       // valid
bar(('' as string | number)); // valid
```

Unifying the signature does work subtly different in this example, just as it does in your example.
I have a feeling that you'll find this for every case the rule enforces.

It's a tradeoff between readability of the code (some people don't like overloads), and strictness of the code.

**Comment by G-Rath** [comment]
That's a fair take - Personally I think that the nature of rules like this imply that they're about letting you know when there isn't a difference (to apply reductio ad absurdum: a `conditional-can-be-simplified` rule shouldn't suggest `a` as a simplification of `a || b`), but I understand it's a tradeoff.

For the example you've given, it's actually [recommended to use a union type](https://www.typescriptlang.org/docs/handbook/declaration-files/do-s-and-don-ts.html#use-union-types), b/c otherwise it'll error like you've shown.

My argument here for why `undefined`/optional is the exception is that it's representing a semantic difference based on if the parameter is passed or not; vs in your example you still always have to pass that parameter, with the parameters type not actually being relevant - hence, a union is "the same".

----

I'm happy to attempt a PR if you'd be happy to accept it, and totally agree it's small fish 🙂

**Comment by bradzacher** [comment]
I'm happy to accept a PR if you want to add this functionality behind a flag!

**Comment by JoshuaKGoldberg** [comment]
Coming back to this: it's been almost 5 years with no other positive interactions for a very niche edge case. I'm going to close this one out. Someone please yell at me if they think that's wrong. 🙂
